### ISSUE TITLE
early return variables

### ISSUE THREAD

**Issue by aceArt-GmbH** [rust-lang/rust-clippy#8377]
### What it does

detect variables that are defined/initialized but not used before an early return

### Lint Name

early-variable

### Category

perf

### Advantage

- Safes computation time
- Maybe safes variable initialization?

### Drawbacks

Users might like to keep the position of their variable declaration

### Example

```rust
fn main() {
    let x = 5;

    if let Some(arg) = env::args().nth(1) {
        match arg.as_str() {
            "-h" | "--help" => {
                return;
            }
            _ => {}
        }
    }

    dbg!(&x);
}
```

Could be written as:

```rust
fn main() {
    if let Some(arg) = env::args().nth(1) {
        match arg.as_str() {
            "-h" | "--help" => {
                return;
            }
            _ => {}
        }
    }

    let x = 5;

    dbg!(&x);
}
```

**Comment by matthiaskrgr** [comment]
Sound more like an llvm optimization to me, hmm.
Maybe this: https://llvm.org/docs/Passes.html#sink-code-sinking

**Comment by aceArt-GmbH** [comment]
Yes, exactly.   
I am not sure if there is much advantage in changing the code if llvm does the optimization anyway. So if this flag is enabled for Rust by default, I guess this issue can be closed then

**Comment by aceArt-GmbH** [comment]
@matthiaskrgr do you know if `-sink` is used by/in rust (release) builds?
How/where could I check it?
